problem1/problem1.cpp

```diff
@@ -0,0 +1,37 @@
+class Foo {
+private:
+    mutex m;
+    condition_variable cv;
+    int turn;
+    
+public:
+    Foo() : turn(0) { }
+
+    void first(function<void()> printFirst) {
+        turnWrapper(printFirst, 0);
+    }
+
+    void second(function<void()> printSecond) {
+        turnWrapper(printSecond, 1);
+    }
+
+    void third(function<void()> printThird) {
+        turnWrapper(printThird, 2);
+    }
+
+    void turnWrapper(function<void()> whichPrint, int whichTurn) {
+        unique_lock<mutex> lock(m);
+        cv.wait(lock, [this, whichTurn]{return turn == whichTurn;});
+
+        try {
+            whichPrint();
+            fflush(stdout); // Woe to he who uses std::cout in a multithreaded environment.
+        } catch (...) {
```

Comment:
ops,  blindly catching and silently ignoring errors can't be good. that's a big red flag - this shouldn't be in production code.

At the same time, this is the only solution which explicitely acknowledges exception safety and attempts to address it. 

Generic advices:
1) consider using try .. finally. instead of catch to, ensure proper release of locks and restoration of invariants
2) you might not even need to do this - given that unique_lock destructor will be automatically invoked on exception
3) it's frequently sufficient to address exception safety by _documenting_ assumptions in the code as well as  behavior of the code you are writing in the presence of exceptions